include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,387 @@
+inline
+void
+TensorProductMatrix<dim,Number,size>
+::vmult(Vector<Number> &dst,
+        const Vector<Number> &src) const
+{
+  AssertDimension(dst.size(), Utilities::fixed_power<dim>(eigenvalues.size()));
+  AssertDimension(src.size(), Utilities::fixed_power<dim>(eigenvalues.size()));
+  vmult(dst.begin(), src.begin());
+}
+
+
+
+template <int dim, typename Number, int size>
+inline
+void
+TensorProductMatrix<dim,Number,size>
+::apply_inverse(Vector<Number> &dst,
+                const Vector<Number> &src) const
+{
+  AssertDimension(dst.size(), Utilities::fixed_power<dim>(eigenvalues.size()));
+  AssertDimension(src.size(), Utilities::fixed_power<dim>(eigenvalues.size()));
+  apply_inverse(dst.begin(), src.begin());
+}
+
+
+
+template <int dim, typename Number, int size>
+inline
+void
+TensorProductMatrix<dim,Number,size>
+::vmult(Number *dst,
+        const Number *src) const
+{
+  const unsigned int n = Utilities::fixed_power<dim>(size > 0 ? size : eigenvalues.size());
+  tmp_array.resize_fast(n*2);
+  const int kernel_size = size > 0 ? size-1 : -1;
+  internal::EvaluatorTensorProduct<internal::evaluate_general,dim,kernel_size,kernel_size+1,Number>
+  eval(AlignedVector<Number>(), AlignedVector<Number>(),
+       AlignedVector<Number>(), mass_matrix.m()-1, mass_matrix.m());
+  const Number *A = &derivative_matrix(0,0);
+  const Number *M = &mass_matrix(0,0);
+  Number *t = tmp_array.begin();
+  if (dim == 1)
+    eval.template apply<0, true, false>(A, src, dst);
+  else if (dim == 2)
+    {
+      eval.template apply<0, true, false>(M, src, t);
+      eval.template apply<1, true, false>(A, t, dst);
+      eval.template apply<0, true, false>(A, src, t);
+      eval.template apply<1, true, true> (M, t, dst);
+    }
+  else if (dim == 3)
+    {
+      eval.template apply<0, true, false>(M, src, t+n);
+      eval.template apply<1, true, false>(M, t+n, t);
+      eval.template apply<2, true, false>(A, t, dst);
+      eval.template apply<1, true, false>(A, t+n, t);
+      eval.template apply<0, true, false>(A, src, t+n);
+      eval.template apply<1, true, true> (M, t+n, t);
+      eval.template apply<2, true, true> (M, t, dst);
+    }
+  else
+    AssertThrow(false, ExcNotImplemented());
+}
+
+
+
+template <int dim, typename Number, int size>
+inline
+void
+TensorProductMatrix<dim,Number,size>
+::apply_inverse(Number *dst,
+                const Number *src) const
+{
+  const unsigned int n = size > 0 ? size : eigenvalues.size();
+  tmp_array.resize_fast(Utilities::fixed_power<dim>(n));
+  const int kernel_size = size > 0 ? size-1 : -1;
+  internal::EvaluatorTensorProduct<internal::evaluate_general,dim,kernel_size,kernel_size+1,Number>
+  eval(AlignedVector<Number>(), AlignedVector<Number>(),
+       AlignedVector<Number>(), mass_matrix.m()-1, mass_matrix.m());
+  const Number *S = &eigenvectors(0,0);
+  Number *t = tmp_array.begin();
+
+  eval.template apply<0, true, false> (S, src, t);
+  if (dim == 1)
```

Comment:
I tried to minimize the number of times I write the `apply` function, but I agree that having nicely separated code is better. (At least until we implement a template recursion functionality for dimension-independent tensor products.)